bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeHandler.java

```diff
@@ -0,0 +1,549 @@
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.gree.internal.handler;
+
+import static org.openhab.binding.gree.internal.GreeBindingConstants.*;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.net.DatagramSocket;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.measure.Unit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.library.unit.ImperialUnits;
+import org.eclipse.smarthome.core.library.unit.SIUnits;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.gree.internal.GreeConfiguration;
+import org.openhab.binding.gree.internal.GreeException;
+import org.openhab.binding.gree.internal.GreeTranslationProvider;
+import org.openhab.binding.gree.internal.discovery.GreeDeviceFinder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link GreeHandler} is responsible for handling commands, which are sent to one of the channels.
+ *
+ * @author John Cunha - Initial contribution
+ * @author Markus Michels - Refactoring, adapted to OH 2.5x
+ */
+@NonNullByDefault
+public class GreeHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(GreeHandler.class);
+    private final GreeTranslationProvider messages;
+    private GreeConfiguration config = new GreeConfiguration();
+    private GreeDeviceFinder deviceFinder = new GreeDeviceFinder();
+    private GreeAirDevice device = new GreeAirDevice();
+    private Optional<DatagramSocket> clientSocket = Optional.empty();
+    private final String thingId;
+    private boolean forceRefresh = false;
+
+    private @Nullable ScheduledFuture<?> refreshTask;
+    private long lastRefreshTime = 0;
+
+    public GreeHandler(Thing thing, GreeTranslationProvider messages, GreeDeviceFinder deviceFinder) {
+        super(thing);
+        this.messages = messages;
+        this.deviceFinder = deviceFinder;
+        String label = getThing().getLabel();
+        this.thingId = label != null ? label : getThing().getUID().getId();
```

Comment:
Thing labels can be changed at runtime, so you should either retrieve the label at runtime at the time when you need to use it, or you should just stick to the thing uid that doesn't change.

Reply:
changed